File: crates/state/src/potoken.js
// Mints a YouTube proof-of-origin token, inside a real browser engine, and leaves it in a cookie.
//
// This is the flow the YouTube web player runs itself. The page publishes a BotGuard challenge and
// its own configuration; the virtual machine behind that challenge attests the browser it is
// running in, Google trades the attestation for an integrity token, and the token yields a minter.
// Only a real engine passes the attestation, which is the whole reason this runs here rather than
// in the app's own javascript interpreter.
//
// The binding to mint for arrives in the query string and the answer goes back in a cookie,
// because reading cookies is all the host can do with the page.
(function () {
  "use strict";

  var COOKIE = "SONORA_POT";
  // What the host reads as a failure rather than a token. A token is base64url and never starts
  // with one of these.
  var FAILED = "!";
  // How long the challenge, and then the virtual machine, are waited for.
  var PATIENCE = 20000;
  var STEP = 100;

  function answer(value) {
    document.cookie = COOKIE + "=" + value + "; path=/; max-age=120";
  }

  function fail(why) {
    answer(
      FAILED +
        String(why)
          .slice(0, 180)
          .replace(/[;,\s]+/g, " "),
    );
  }

  // The binding, read the moment this script runs. The page rewrites its own url before it
  // finishes loading and navigates away from it afterwards, so it is read at document start and
  // kept in the tab's own storage for whatever page comes next.
  var wanted = (function () {
    var found = /[?&]binding=([^&]+)/.exec(location.search);
    var value = found ? decodeURIComponent(found[1]) : null;
    try {
      if (value) sessionStorage.setItem(COOKIE, value);
      else value = sessionStorage.getItem(COOKIE);
    } catch (error) {
      // A page that refuses storage still works on its first load.
    }
    return value;
  })();

  // Whether a page before this one already left a token. Only a failure is worth replacing.
  function minted() {
    var found = new RegExp("(?:^|; )" + COOKIE + "=([^;]*)").exec(
      document.cookie,
    );
    return found && found[1] && found[1].charAt(0) !== FAILED;
  }

  // Waits for `condition` to return something, checking every `STEP` until `PATIENCE` is up.
  function until(condition, why) {
    return new Promise(function (resolve, reject) {
      var waited = 0;
      (function look() {
        var found = condition();
        if (found) return resolve(found);
        if ((waited += STEP) > PATIENCE) return reject(new Error(why));
        setTimeout(look, STEP);
      })();
    });
  }

  // The challenge, taken as the page hands it over.
  //
  // YouTube publishes it by calling `window.ytAtN`, a resolver it creates for its own promise and
  // deletes on `DOMContentLoaded`, passing nothing when the page carries no challenge. Standing
  // between the two is the only way to see the payload whichever order those happen in: the page's
  // resolver goes in through the setter, and everyone who reads `window.ytAtN` gets a wrapper that
  // records the payload and passes it straight on.
  var captured = null;
  var resolver = null;

  function relay(payload) {
    if (payload) captured = payload;
    if (resolver) resolver(payload);
  }

  try {
    Object.defineProperty(window, "ytAtN", {
      configurable: true,
      get: function () {
        return resolver ? relay : undefined;
      },
      set: function (value) {
        resolver = value;
      },
    });
  } catch (error) {
    // A page that will not take the property still loads; it just never mints.
  }

  function challenge() {
    return until(function () {
      return captured;
    }, "the page published no botguard challenge").then(function (payload) {
      var response = payload.R;
      if (typeof response === "string") response = JSON.parse(response);
      var found = response && response.bgChallenge;
      if (!found || !found.program || !found.interpreterUrl) {
        throw new Error("the challenge carries no program");
      }
      return found;
    });
  }

  function base64ToBytes(value) {
    var binary = atob(value.replace(/-/g, "+").replace(/_/g, "/"));
    var bytes = new Uint8Array(binary.length);
    for (var at = 0; at < binary.length; at++)
      bytes[at] = binary.charCodeAt(at);
    return bytes;
  }

  function bytesToBase64(bytes) {
    var binary = "";
    for (var at = 0; at < bytes.length; at++)
      binary += String.fromCharCode(bytes[at]);
    return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_");
  }

  // The page's trusted types policy, which its own scripts go through. Youtube refuses a plain
  // string for a script url, so one has to be minted here too; a page that will not hand out a
  // policy leaves `source` returning the string and the assignment below fails loudly.
  var policy = (function () {
    if (!window.trustedTypes || !window.trustedTypes.createPolicy) return null;
    try {
      return window.trustedTypes.createPolicy("sonora-potoken", {
        createScriptURL: function (url) {
          return url;
        },
      });
    } catch (error) {
      return null;
    }
  })();

  function source(url) {
    return policy ? policy.createScriptURL(url) : url;
  }

  // Brings the interpreter in the way the page's own player does, as a script element. Fetching it
  // instead would need cross-origin permission google.com does not give, and running the source
  // through `new Function` would need the eval the page's content policy forbids.
  function load(url) {
    return new Promise(function (resolve, reject) {
      var element = document.createElement("script");
      element.onload = function () {
        resolve();
      };
      element.onerror = function () {
        reject(new Error("cannot load the botguard interpreter"));
      };
      try {
        element.src = source(url);
      } catch (error) {
        return reject(
          new Error("the page refused the interpreter url: " + error.message),
        );
      }
      document.head.appendChild(element);
    });
  }

  // Loads the virtual machine and takes one snapshot of it. The third element of the array handed
  // to the snapshot is where BotGuard leaves the factory that makes a minter, which is the part we
  // are here for; an engine it does not trust gets a snapshot and an empty array.
  function attest(bg) {
    var vm = window[bg.globalName];
    if (!vm || !vm.a)
      return Promise.reject(new Error("the interpreter defined no vm"));
    var functions = null;
    try {
      vm.a(
        bg.program,
        function (snapshot, shutdown, passEvent, checkCamera) {
          functions = {
            snapshot: snapshot,
            shutdown: shutdown,
            passEvent: passEvent,
            checkCamera: checkCamera,
          };
        },
        true,
        undefined,
        function () {},
        [[], []],
        undefined,
        false,
        undefined,
      );
    } catch (error) {
      return Promise.reject(new Error("the vm refused its program"));
    }
    return until(function () {
      return functions;
    }, "the vm never handed back its functions").then(function (ready) {
      return new Promise(function (resolve, reject) {
        var signals = [];
        var settled = false;
        setTimeout(function () {
          if (!settled) reject(new Error("the snapshot timed out"));
        }, PATIENCE);
        ready.snapshot(
          function (token) {
            settled = true;
            resolve({ token: token, signals: signals });
          },
          [undefined, undefined, signals, undefined],
        );
      });
    });
  }

  // The player script the page is running, which is where the attestation constants live. Its url
  // is in the page's own configuration, and it is same-origin, so the browser usually answers this
  // out of the cache it filled when the page loaded.
  function player() {
    var url =
      (window.yt && window.yt.config_ && window.yt.config_.PLAYER_JS_URL) ||
      (window.ytcfg && window.ytcfg.get && window.ytcfg.get("PLAYER_JS_URL"));
    if (!url) {
      var found = /"jsUrl"\s*:\s*"([^"]+base\.js)"/.exec(
        document.documentElement.innerHTML,
      );
      url = found && found[1];
    }
    if (!url)
      return Promise.reject(new Error("the page names no player script"));
    return fetch(url).then(function (response) {
      return response.text();
    });
  }

  // The attestation constants, read out of the player script.
  //
  // Nothing here has a hardcoded fallback on purpose. A pattern only misses when Google has moved
  // the thing it was looking for, which is exactly when a remembered value is stale, so guessing
  // would trade a truthful failure for a request that is refused for a reason the log then lies
  // about. The mint is not on the playback path and a cold start token covers the gap, so failing
  // and saying which piece went missing costs one retry and buys a name to go and fix.
  function constants() {
    return player().then(function (source) {
      // The player reads its key off an experiment and keeps two literals behind it, of which the
      // web client takes the second. A third branch, or a swapped order, would need reading again.
      var branches = /html5_web_po_request_key[\s\S]*?\}/.exec(source);
      var literals = branches ? branches[0].match(/"[A-Za-z0-9_-]{20}"/g) : null;
      var key = /"X-Goog-Api-Key"\]\s*:\s*"(AIzaSy[A-Za-z0-9_-]{33})"/.exec(source);
      var host = /"(https:\/\/[a-z0-9-]+-pa\.googleapis\.com)"/.exec(source);
      var path = /"(\/google\.internal\.waa\.v1\.Waa\/GenerateIT)"/.exec(source);
      var missing = [];
      if (!literals) missing.push("request key");
      if (!key) missing.push("api key");
      if (!host) missing.push("anti-abuse host");
      if (!path) missing.push("rpc path");
      if (missing.length) {
        throw new Error("the player names no " + missing.join(" and no "));
      }
      return {
        request: literals[literals.length - 1].slice(1, -1),
        key: key[1],
        host: host[1],
        path: path[1],
      };
    });
  }

  // Trades an attestation for an integrity token. Google answers a browser it does not trust with
  // a fallback token and no integrity token at all.
  function integrity(attestation, waa) {
    return fetch(waa.host + "/$rpc" + waa.path, {
      method: "POST",
      headers: {
        "content-type": "application/json+protobuf",
        "x-goog-api-key": waa.key,
        "x-user-agent": "grpc-web-javascript/0.1",
      },
      body: JSON.stringify([waa.request, attestation]),
    })
      .then(function (response) {
        return response.json();
      })
      .then(function (issued) {
        if (!issued || !issued[0])
          throw new Error("google issued no integrity token");
        return base64ToBytes(issued[0]);
      });
  }

  function mint() {
    if (!wanted) return fail("no binding in the url");
    if (minted()) return;
    var signals = null;
    challenge()
      .then(function (bg) {
        var url =
          bg.interpreterUrl
            .privateDoNotAccessOrElseTrustedResourceUrlWrappedValue;
        return load("https:" + url).then(function () {
          return attest(bg);
        });
      })
      .then(function (attested) {
        if (!attested.signals[0])
          throw new Error("botguard did not trust this browser");
        signals = attested.signals;
        return constants().then(function (waa) {
          return integrity(attested.token, waa);
        });
      })
      .then(function (token) {
        return signals[0](token);
      })
      .then(function (minter) {
        if (typeof minter !== "function")
          throw new Error("the integrity token yielded no minter");
        return minter(new TextEncoder().encode(wanted));
      })
      .then(function (token) {
        if (!token || !token.length)
          throw new Error("the minter returned nothing");
        answer(bytesToBase64(token));
      })
      .catch(function (error) {
        fail(error && error.message ? error.message : error);
      });
  }

  // Nothing waits for the load event. A youtube page in a window no one is looking at can keep
  // loading for a long time, and the challenge arrives well before it settles; `challenge` waits
  // for that instead.
  try {
    mint();
  } catch (error) {
    fail(error && error.message ? error.message : error);
  }
})();

File: crates/state/src/potoken.rs
//! Minting YouTube's proof-of-origin token in a hidden browser window.
//!
//! YouTube stops serving streams to an address it has flagged unless the request carries a token
//! that proves it came from a browser. Google's BotGuard virtual machine produces one only after
//! attesting the engine it is running in, and no hand-written javascript environment passes that,
//! so the app opens the one real engine it ships: a `webview` page on youtube.com running
//! `potoken.js`, off the screen, which leaves the token in a cookie.
//!
//! Nothing here drives it on a schedule. `music::potoken` records the identifier the YouTube
//! client is waiting on, this entity notices, mints once, and files the answer; the client uses a
//! cold start token until then, which covers the start of a track and no more. A platform with no
//! webview, the Flatpak runtime among them, never opens a window and never mints.

use std::time::{Duration, Instant};

use gpui::{App, AppContext as _, Context, Task};

/// Where the minting page is loaded from. The binding rides in the query string because reading
/// cookies is the only channel back out of the page.
const PAGE: &str = "https://www.youtube.com/";
const LANDING: &str = "www.youtube.com";
const DOMAIN: &str = "youtube.com";
/// The cookie the page leaves its answer in.
const COOKIE: &str = "SONORA_POT";
/// What the page prefixes a failure with. A token is base64url and never starts with one.
const FAILED: char = '!';
/// How often the entity looks for an ask and polls a window it has open.
const TICK: Duration = Duration::from_millis(500);
/// How long a page gets to attest and mint before it is given up on. Google's round trip takes a
/// second or two; a page still quiet after this is not going to answer.
const PATIENCE: Duration = Duration::from_secs(45);

/// The browser window that mints proof-of-origin tokens, and the one ask it is serving.
pub struct PoToken {
    page: Option<webview::Page>,
    /// What the open page is minting for.
    binding: Option<String>,
    opened: Option<Instant>,
    task: Option<Task<()>>,
}

impl PoToken {
    pub fn new(cx: &mut Context<Self>) -> Self {
        let mut this = Self {
            page: None,
            binding: None,
            opened: None,
            task: None,
        };
        this.watch(cx);
        this
    }

    /// Ticks for the life of the app: one tick starts a mint, carries one along, or does nothing.
    fn watch(&mut self, cx: &mut Context<Self>) {
        self.task = Some(cx.spawn(async move |this, cx| {
            loop {
                cx.background_executor().timer(TICK).await;
                if this.update(cx, |this, cx| this.tick(cx)).is_err() {
                    return;
                }
            }
        }));
    }

    fn tick(&mut self, cx: &mut Context<Self>) {
        match self.page.is_some() {
            true => self.carry(),
            false => self.start(cx),
        }
    }

    /// Opens a page for whatever the client is waiting on, if anything.
    fn start(&mut self, _cx: &mut Context<Self>) {
        let Some(binding) = music::potoken::wanted() else {
            return;
        };
        if !webview::supported() {
            log::debug!(
                "potoken: no browser engine here, the stream goes out on a cold start token"
            );
            music::potoken::failed(&binding);
            return;
        }
        let target = webview::Target {
            url: format!("{PAGE}?binding={}", escaped(&binding)),
            landing: LANDING.to_string(),
            domain: DOMAIN.to_string(),
            proof: vec![COOKIE.to_string()],
            title: String::new(),
            agent: None,
            script: Some(include_str!("potoken.js").to_string()),
        };
        match webview::Page::open(target) {
            Ok(page) => {
                log::debug!("potoken: minting for {binding}");
                self.page = Some(page);
                self.binding = Some(binding);
                self.opened = Some(Instant::now());
            }
            Err(error) => {
                log::warn!("potoken: cannot open the minting page: {error:#}");
                music::potoken::failed(&binding);
            }
        }
    }

    /// Carries the open page one poll further, and files whatever it produced.
    fn carry(&mut self) {
        let Some(page) = self.page.as_mut() else {
            return;
        };
        let polled = page.poll();
        let binding = self.binding.clone().unwrap_or_default();
        match polled {
            webview::Poll::Pending => {
                if self.overdue() {
                    log::warn!("potoken: the minting page did not answer in {PATIENCE:?}");
                    self.done(&binding, None);
                }
            }
            webview::Poll::Closed => {
                log::warn!("potoken: the minting page closed before it answered");
                self.done(&binding, None);
            }
            webview::Poll::Cookies(header) => {
                let answer = value(&header, COOKIE).unwrap_or_default();
                match answer.strip_prefix(FAILED) {
                    Some(why) => {
                        log::warn!("potoken: the page could not mint a token: {why}");
                        self.done(&binding, None);
                    }
                    None => {
                        log::debug!("potoken: minted a token for {binding}");
                        self.done(&binding, Some(answer.to_string()));
                    }
                }
            }
        }
    }

    fn overdue(&self) -> bool {
        self.opened.is_some_and(|at| at.elapsed() > PATIENCE)
    }

    /// Files the outcome and closes the page. A failure is filed too, which is what holds the next
    /// attempt off rather than reopening a window every tick.
    fn done(&mut self, binding: &str, token: Option<String>) {
        self.page = None;
        self.binding = None;
        self.opened = None;
        match token {
            Some(token) => music::potoken::give(binding, token),
            None => music::potoken::failed(binding),
        }
    }
}

/// Reads one cookie out of a `Cookie` header value.
fn value<'a>(header: &'a str, name: &str) -> Option<&'a str> {
    header
        .split(';')
        .filter_map(|pair| pair.trim().split_once('='))
        .find(|(key, _)| *key == name)
        .map(|(_, value)| value)
}

/// Percent-encodes everything a query value may not carry. Visitor data arrives base64url and a
/// data sync id carries `||`, so neither is safe to paste into a url as it stands.
fn escaped(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for byte in value.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'.' | b'_' | b'~' => {
                out.push(byte as char)
            }
            _ => out.push_str(&format!("%{byte:02X}")),
        }
    }
    out
}

/// Starts the minting window. Nothing holds it but the global, and nothing reads it: the client
/// and the window meet in `music::potoken`.
pub fn attach(cx: &mut App) -> gpui::Entity<PoToken> {
    cx.new(PoToken::new)
}
